f2py.compile issues (#7683)

bertrand-l opened the pull request:
- DOC: add to `f2py.compile`'s doctstring a description of the `source_fn` argument
- MAINT: avoid warnings about deprecated `imp.load_module` in `distutil.misc_util` with Python 3

That leaves one issue in #7683 that I have not yet figured out:

```
Could not locate executable C:ProgramsPython35python.exe
Executable C:ProgramsPython35python.exe does not exist
```

Looks like some string escaping issue, but nothing jumped at me looking at `distutil.exec_command.{exec_command,find_executable}` and I don't have a windows machine to test on.

bertrand-l committed:
MAINT,DOC: add to compat.py3k a function to load modules. Fix some doc for f2py.compile (issue #7683)

charris commented on numpy/f2py/__init__.py:
```diff
@@ -35,8 +35,11 @@ def compile(source,
     verbose: bool, optional
         print f2py output to screen
+    source_fn : str, optional
+        filename where the fortran source is dumped
```
Could use a bit more explanation. I can't figure out what "dumped" means in this context ;)

bertrand-l replied:
I see your point :)
```
    source_fn : str, optional
        name of the file where the fortran source is written
        before being compiled. The default is to use a temporary
        file with an extension provided by the `extension` parameter
```

charris commented on numpy/distutils/misc_util.py:
```diff
@@ -2289,6 +2286,20 @@ def msvc_version(compiler):
+    def import_module(name, fn, info=('.py', 'U', 1)):
+        """Import and return module"""
+        import imp
+        fo = open(fn, info[1])
```
Could use context manager here.
```
with open(fn, info[0]) as fo:
    return mp.load_module(name, fo, fn, info)
```

bertrand-l replied:
Thanks, this was just copy/paste of the old code and somehow i missed that :|

charris commented on numpy/distutils/misc_util.py:
```diff
@@ -1912,11 +1909,11 @@ def get_version(self, version_file=None, version_variable=None):
                 n = dot_join(self.name, name)
                 try:
-                    version_module = imp.load_module('_'.join(n.split('.')),*info)
+                    version_module = import_module('_'.join(n.split('.')), *info)
```
`*info` looks wrong, should omit the `*` I think.

charris replied:
NVM. Hmm, I think it would be clearer change info to `('.py', 'U', 1)`, then
```
import_module('_'.join(n.split('.')), fn, info)
```

charris replied:
I haven't checked that the old form of `info` is needed elsewhere.

charris replied:
The change looks safe. The line is also too long, so should break it.

bertrand-l replied:
Agreed, that old form of `info` is a bit confusing (and it does not seem to be used anywhere else)

charris commented on numpy/distutils/misc_util.py:
```diff
@@ -2289,6 +2286,20 @@ def msvc_version(compiler):
+        """Import and return module"""
+        import imp
+        fo = open(fn, info[1])
+        module = imp.load_module(name, fo, fn, info)
```
And here is should be `*info`.

charris commented:
Looks generally good. Could you also fix the other `imp` uses? there is one in `numpy/_import_tools.py` that could use the same fix as here, and another in `numpy/testing/utils.py`. Hmm, might make sense to put your `import_module` function in `numpy/compat/py3k.py`

bertrand-l commented on numpy/f2py/__init__.py:
```diff
@@ -35,8 +35,11 @@ def compile(source,
     extension: {'.f', '.f90'}, optional
-        filename extension influences the fortran compiler behavior
+        filename extension if `source_fn` is not provided.
+        Extension influences the fortran compiler behavior
```
```
    extension: {'.f', '.f90'}, optional
        filename extension if `source_fn` is not provided.
        The extension tells the compiler which fortran standard
        to use. The default is `.f`, which implies F77 standard.
```

bertrand-l commented:
Good point. Will do and then update the PR. Thanks for the review!

charris commented:
If you put it in `py3k.py`, `npy_import_module` might be a better name.

bertrand-l commented:
@charris I think i've addressed all your comments (apart from the context manager in `npy_import_module` - keeping `try: ... finally: ...` block because `imp.find_module` returns a file object). Thanks again!

charris commented on numpy/compat/py3k.py:
```diff
@@ -91,9 +91,47 @@ def asunicode_nested(x):
+        """
+        Import and return module
+
+        name : str
```
Needs the heading
```
Parameters
-----------------
```

See `doc/HOWTO_DOCUMENT.rst.txt` or other functions for docstring.

bertrand-l replied:
Done.

charris commented on numpy/compat/py3k.py:
```diff
@@ -91,9 +91,47 @@ def asunicode_nested(x):
+        """
+        Import and return module
+
+        name : str
```
See above about docstring.

charris commented on numpy/f2py/__init__.py:
```diff
@@ -30,13 +30,19 @@ def compile(source,
         Fortran source of module / subroutine to compile
     modulename : str, optional
-        the name of compiled python module
+        name of the compiled python module, 'untitled' by default
```
Should capitalize the first word in a sentence. Keeping the `The` is better.

bertrand-l replied:
Done.

charris commented on numpy/f2py/__init__.py:
```diff
@@ -30,13 +30,19 @@ def compile(source,
+        additional parameters passed to f2py, empty by default
     verbose: bool, optional
-        print f2py output to screen
+        print f2py output to screen, `True` by default
```
See above about capitalization. Same on down for other variable documentation.

bertrand-l replied:
Done (as well as removing the defaults which are explicit in the signature)

charris commented on numpy/f2py/__init__.py:
```diff
@@ -30,13 +30,19 @@ def compile(source,
     modulename : str, optional
-        the name of compiled python module
+        name of the compiled python module, 'untitled' by default
     extra_args: str, optional
```
Should have space before `:`. Same below.

bertrand-l replied:
Done.

charris commented:
Couple of style nits. Also, please squash the commits using `numpy rebase -i HEAD~4`. You can reorder the commits without problems (cut and paste) as long as they don't touch the same files, then squash them by replacing `pick` with `s`.

bertrand-l commented:
All done, including the squashing. Also renamed `compat.py3k.npy_import_module` to `compat.py3k.npy_load_module` as it is closer to the name of the original python functions.

bertrand-l commented on numpy/_import_tools.py:
```diff
@@ -86,10 +86,9 @@ def _init_info_modules(self, packages=None):
-                                              filedescriptor)
+                info_module = npy_load_module(fullname + '.info',
+                                                info_file,
+                                                filedescriptor)
```
oops, will fix that indentation, sorry

charris commented:
The new `npy_load_module` function needs `.. versionadded:: 1.11.12` at the end of the summary, grep for examples. LGTM otherwise.

bertrand-l commented:
ok thanks (and sorry there was yet one more thing i forgot). `npy_load_module` now has a `versionadded`.

seberg commented on numpy/compat/py3k.py:
```diff
@@ -91,9 +91,66 @@ def asunicode_nested(x):
+        """
+        Load a module.
+
+        .. versionadded:: 1.11.12
```
Typo, I don't think we are 12 yet ;), seems to be 1.11.2

bertrand-l commented:
@seberg, thanks. The `versionadded` is fixed now.

charris commented:
Thanks @bertrand-l .

charris commented:
For next time, note that your commit message summary line is too long. Lines should be kept <= 72 characters.

bertrand-l commented:
Thanks @charris for the review and merge (and sorry it took so many iterations and comments)
